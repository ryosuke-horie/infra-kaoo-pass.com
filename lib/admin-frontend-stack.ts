import {
  RemovalPolicy,
  Stack,
  type StackProps,
  aws_cloudfront,
  aws_cloudfront_origins,
  aws_iam,
  aws_s3,
} from "aws-cdk-lib";
import type { Construct } from "constructs";

import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import * as route53 from "aws-cdk-lib/aws-route53";
import * as targets from "aws-cdk-lib/aws-route53-targets";

/**
 * CloudFront + S3でホスティング用のStackを作成する
 */
export class AdminFrontendStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // S3バケットを作成する
    const adminWebsiteBucket = new aws_s3.Bucket(this, "adminWebsiteBucket", {
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // OAIを作成する
    const originAccessIdentity = new aws_cloudfront.OriginAccessIdentity(
      this,
      "OriginAccessIdentity",
      {
        comment: "admin-website-distribution-originAccessIdentity",
      }
    );

    // S3バケットポリシーを作成する。OAIからのみアクセス可能とする
    const adminWebsiteBucketPolicyStatement = new aws_iam.PolicyStatement({
      // GETのみ許可
      actions: ["s3:GetObject"],
      effect: aws_iam.Effect.ALLOW,
      principals: [
        new aws_iam.CanonicalUserPrincipal(
          originAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId
        ),
      ],
      resources: [`${adminWebsiteBucket.bucketArn}/*`],
    });

    // S3バケットポリシーにステートメントを追加する
    adminWebsiteBucket.addToResourcePolicy(adminWebsiteBucketPolicyStatement);

    // 利用するホストゾーンをドメイン名で取得
    // ホストゾーンIDを取得
    const hostedZoneId = route53.HostedZone.fromLookup(this, "HostedZoneId", {
      domainName: "kaoo-pass.com",
    });

    // 証明書を取得
    const certificate = Certificate.fromCertificateArn(
      this,
      "Certificate",
      "arn:aws:acm:us-east-1:905418074681:certificate/361694cc-b668-454e-83ae-4a4f9641951a"
    );

    // CloudFrontディストリビューションを作成する
    const distribution = new aws_cloudfront.Distribution(this, "Distribution", {
        domainNames: ["admin.kaoo-pass.com"],
        certificate,
      comment: "admin.kaoo-pass.com",
      defaultRootObject: "index.html",
      defaultBehavior: {
        allowedMethods: aws_cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
        cachedMethods: aws_cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
        cachePolicy: aws_cloudfront.CachePolicy.CACHING_OPTIMIZED,
        viewerProtocolPolicy:
          aws_cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        origin: new aws_cloudfront_origins.S3Origin(adminWebsiteBucket, {
          originAccessIdentity,
        }),
      },

      // 403エラーの時に「/」に飛ばす(SPAのため、リロードや直アクセスに対策するため)
      errorResponses: [
        {
          httpStatus: 403,
          responseHttpStatus: 200,
          responsePagePath: "/",
        },
      ],
      priceClass: aws_cloudfront.PriceClass.PRICE_CLASS_100,
    });

    // Route53レコード設定
    new route53.ARecord(this, "ARecord", {
      zone: hostedZoneId,
      target: route53.RecordTarget.fromAlias(
        new targets.CloudFrontTarget(distribution)
      ),
      recordName: "admin.kaoo-pass.com",
    });
  }
}
