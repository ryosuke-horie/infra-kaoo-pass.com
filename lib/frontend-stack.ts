import {
	RemovalPolicy,
	Stack,
	type StackProps,
	aws_cloudfront,
	aws_cloudfront_origins,
	aws_iam,
	aws_lambda,
	aws_s3,
} from "aws-cdk-lib";
import type { Construct } from "constructs";

import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import * as route53 from "aws-cdk-lib/aws-route53";
import * as targets from "aws-cdk-lib/aws-route53-targets";

/**
 * CloudFront + S3でホスティング用のStackを作成する
 */
export class FrontendStack extends Stack {
	constructor(scope: Construct, id: string, props?: StackProps) {
		super(scope, id, props);

		// S3バケットを作成する
		const websiteBucket = new aws_s3.Bucket(this, "WebsiteBucket", {
			removalPolicy: RemovalPolicy.DESTROY,
		});

		// 画像用のS3バケットを作成する
		const imageBucket = new aws_s3.Bucket(this, "ImageBucket", {
			removalPolicy: RemovalPolicy.DESTROY,
		});

		// OAIを作成する
		const originAccessIdentity = new aws_cloudfront.OriginAccessIdentity(
			this,
			"OriginAccessIdentity",
			{
				comment: "website-distribution-originAccessIdentity",
			},
		);

		// OAIを作成する
		const imageOriginAccessIdentity = new aws_cloudfront.OriginAccessIdentity(
			this,
			"ImageOriginAccessIdentity",
			{
				comment: "image-distribution-originAccessIdentity",
			},
		);

		// S3バケットポリシーを作成する。OAIからのみアクセス可能とする
		const websiteBucketPolicyStatement = new aws_iam.PolicyStatement({
			// GETのみ許可
			actions: ["s3:GetObject"],
			effect: aws_iam.Effect.ALLOW,
			principals: [
				new aws_iam.CanonicalUserPrincipal(
					originAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId,
				),
			],
			resources: [`${websiteBucket.bucketArn}/*`],
		});

		// S3バケットポリシーを作成する。OAIからのみアクセス可能とする
		const imageBucketPolicyStatement = new aws_iam.PolicyStatement({
			// GETのみ許可
			actions: ["s3:GetObject"],
			effect: aws_iam.Effect.ALLOW,
			principals: [
				new aws_iam.CanonicalUserPrincipal(
					imageOriginAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId,
				),
			],
			resources: [`${imageBucket.bucketArn}/*`],
		});

		// S3バケットポリシーにステートメントを追加する
		websiteBucket.addToResourcePolicy(websiteBucketPolicyStatement);

		// S3バケットポリシーにステートメントを追加する
		imageBucket.addToResourcePolicy(imageBucketPolicyStatement);

		// // 利用するホストゾーンをドメイン名で取得
		// // ホストゾーンIDを取得
		const hostedZoneId = route53.HostedZone.fromLookup(this, "HostedZoneId", {
			domainName: "kaoo-pass.com",
		});

		// // 証明書を取得
		const certificate = Certificate.fromCertificateArn(
			this,
			"Certificate",
			"arn:aws:acm:us-east-1:905418074681:certificate/01d9e454-018a-4063-96f8-b20ad66ea2c1",
		);

		// CloudFrontディストリビューションを作成する
		const distribution = new aws_cloudfront.Distribution(this, "Distribution", {
			domainNames: ["kaoo-pass.com"],
			certificate,
			comment: "kaoo-pass.com",
			defaultRootObject: "index.html",
			defaultBehavior: {
				allowedMethods: aws_cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
				cachedMethods: aws_cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
				cachePolicy: aws_cloudfront.CachePolicy.CACHING_OPTIMIZED,
				viewerProtocolPolicy:
					aws_cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
				origin: new aws_cloudfront_origins.S3Origin(websiteBucket, {
					originAccessIdentity,
				}),
			},
			// 画像用のビヘイビアを追加する
			additionalBehaviors: {
				"/images/*": {
					origin: new aws_cloudfront_origins.S3Origin(imageBucket, {
						originAccessIdentity: imageOriginAccessIdentity,
					}),
					viewerProtocolPolicy:
						aws_cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
					allowedMethods: aws_cloudfront.AllowedMethods.ALLOW_GET_HEAD,
					cachedMethods: aws_cloudfront.CachedMethods.CACHE_GET_HEAD,
					compress: true,
					originRequestPolicy:
						aws_cloudfront.OriginRequestPolicy.CORS_S3_ORIGIN,
					responseHeadersPolicy:
						aws_cloudfront.ResponseHeadersPolicy.CORS_ALLOW_ALL_ORIGINS,
				},
			},

			// 403エラーの時に「/」に飛ばす(SPAのため、リロードや直アクセスに対策するため)
			errorResponses: [
				{
					httpStatus: 403,
					responseHttpStatus: 200,
					responsePagePath: "/",
				},
			],
			priceClass: aws_cloudfront.PriceClass.PRICE_CLASS_100,
		});

		// // Route53レコード設定
		new route53.ARecord(this, "ARecord", {
			zone: hostedZoneId,
			target: route53.RecordTarget.fromAlias(
				new targets.CloudFrontTarget(distribution),
			),
			recordName: "kaoo-pass.com",
		});

		// 画像用S3へのアクセス権限を追加
		// BackendのAPI用Lambdaを検索する
		const apiLambdaRole = aws_iam.Role.fromRoleArn(
			this,
			"ApiLambdaRole",
			"arn:aws:iam::905418074681:role/laravel-api-dev-ap-northeast-1-lambdaRole",
		);

		// API用LambdaにS3バケットへのアクセス権限を追加する
		imageBucket.grantReadWrite(apiLambdaRole);
	}
}
