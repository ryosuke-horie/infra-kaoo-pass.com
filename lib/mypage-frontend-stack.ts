import {
	RemovalPolicy,
	Stack,
	type StackProps,
	aws_cloudfront,
	aws_cloudfront_origins,
	aws_iam,
	aws_s3,
} from "aws-cdk-lib";
import type { Construct } from "constructs";

import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import * as route53 from "aws-cdk-lib/aws-route53";
import * as targets from "aws-cdk-lib/aws-route53-targets";

/**
 * CloudFront + S3でホスティング用のStackを作成する
 */
export class MypageFrontendStack extends Stack {
	constructor(scope: Construct, id: string, props?: StackProps) {
		super(scope, id, props);

		// S3バケットを作成する
		const mypageWebsiteBucket = new aws_s3.Bucket(this, "mypageWebsiteBucket", {
			removalPolicy: RemovalPolicy.DESTROY,
		});

		// OAIを作成する
		const originAccessIdentity = new aws_cloudfront.OriginAccessIdentity(
			this,
			"OriginAccessIdentity",
			{
				comment: "mypage-website-distribution-originAccessIdentity",
			},
		);

		// S3バケットポリシーを作成する。OAIからのみアクセス可能とする
		const mypageWebsiteBucketPolicyStatement = new aws_iam.PolicyStatement({
			// GETのみ許可
			actions: ["s3:GetObject"],
			effect: aws_iam.Effect.ALLOW,
			principals: [
				new aws_iam.CanonicalUserPrincipal(
					originAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId,
				),
			],
			resources: [`${mypageWebsiteBucket.bucketArn}/*`],
		});

		// S3バケットポリシーにステートメントを追加する
		mypageWebsiteBucket.addToResourcePolicy(mypageWebsiteBucketPolicyStatement);

		// 利用するホストゾーンをドメイン名で取得
		// ホストゾーンIDを取得;
		const hostedZoneId = route53.HostedZone.fromLookup(this, "HostedZoneId", {
			domainName: "kaoo-pass.com",
		});

		// 証明書を取得
		const certificate = Certificate.fromCertificateArn(
			this,
			"Certificate",
			"arn:aws:acm:us-east-1:905418074681:certificate/bfb07cff-ec5d-4b66-af24-2d45086a0ace",
		);

		// CloudFrontディストリビューションを作成する
		const distribution = new aws_cloudfront.Distribution(this, "Distribution", {
			domainNames: ["mypage.kaoo-pass.com"],
			certificate,
			comment: "mypage.kaoo-pass.com",
			defaultRootObject: "index.html",
			defaultBehavior: {
				allowedMethods: aws_cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
				cachedMethods: aws_cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
				cachePolicy: aws_cloudfront.CachePolicy.CACHING_OPTIMIZED,
				viewerProtocolPolicy:
					aws_cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
				origin: new aws_cloudfront_origins.S3Origin(mypageWebsiteBucket, {
					originAccessIdentity,
				}),
			},

			// 403エラーの時に「/」に飛ばす(SPAのため、リロードや直アクセスに対策するため)
			errorResponses: [
				{
					httpStatus: 403,
					responseHttpStatus: 200,
					responsePagePath: "/",
				},
			],
			priceClass: aws_cloudfront.PriceClass.PRICE_CLASS_100,
		});

		// Route53レコード設定
		new route53.ARecord(this, "ARecord", {
			zone: hostedZoneId,
			target: route53.RecordTarget.fromAlias(
				new targets.CloudFrontTarget(distribution),
			),
			recordName: "mypage.kaoo-pass.com",
		});
	}
}
