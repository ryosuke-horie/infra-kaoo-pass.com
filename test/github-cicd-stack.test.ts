import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { describe, expect, it } from "vitest";
import { GithubCiCdStack } from "../lib/github-cicd-stack";

describe("GithubCiCdStack", () => {
	it("should create the correct resources", () => {
		const app = new cdk.App();
		const stack = new GithubCiCdStack(app, "TestGithubCiCdStack", {
			env: { region: "ap-northeast-1", account: "905418074681" },
		});
		const template = Template.fromStack(stack);

		expect(template.toJSON()).toMatchSnapshot();
	});
});
