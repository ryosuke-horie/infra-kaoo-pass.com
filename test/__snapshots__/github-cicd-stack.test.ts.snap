// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GithubCiCdStack > should create the correct resources 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CdkDeployPolicy4C829943": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:getBucketLocation",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::*",
            },
            {
              "Action": [
                "cloudformation:CreateStack",
                "cloudformation:CreateChangeSet",
                "cloudformation:DeleteChangeSet",
                "cloudformation:DescribeChangeSet",
                "cloudformation:DescribeStacks",
                "cloudformation:DescribeStackEvents",
                "cloudformation:ExecuteChangeSet",
                "cloudformation:GetTemplate",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:cloudformation:ap-northeast-1:905418074681:stack/*/*",
            },
            {
              "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::frontendstack-websitebucket75c24d94-gjegy088wi5q/*",
            },
            {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:ap-northeast-1:905418074681:parameter/cdk-bootstrap/hnb659fds/version",
            },
            {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::905418074681:role/cdk-hnb659fds-cfn-exec-role-905418074681-ap-northeast-1",
            },
            {
              "Action": [
                "cloudfront:GetDistribution",
                "cloudfront:GetDistributionConfig",
                "cloudfront:ListDistributions",
                "cloudfront:ListStreamingDistributions",
                "cloudfront:CreateInvalidation",
                "cloudfront:ListInvalidations",
                "cloudfront:GetInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:cloudfront::905418074681:distribution/E7STCAA1QIEGW",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkDeployPolicy",
        "Roles": [
          {
            "Ref": "GitHubOidcRoleD5E95968",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": {
      "DependsOn": [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-905418074681-ap-northeast-1",
          "S3Key": "977fc1649d2dbcce16e23f6332faef6fa0f48aa74a0afe35f4a3467754e20cd8.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GitHubOidcProvider7EBF861F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ClientIDList": [
          "sts.amazonaws.com",
        ],
        "CodeHash": "977fc1649d2dbcce16e23f6332faef6fa0f48aa74a0afe35f4a3467754e20cd8",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "GitHubOidcRoleD5E95968": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                  "token.actions.githubusercontent.com:sub": [
                    "repo:ryosuke-horie/mem-vision-front:ref:refs/heads/main",
                    "repo:ryosuke-horie/mem-vision-serverless:ref:refs/heads/main",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "GitHubOidcProvider7EBF861F",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "GitHubOidcRole",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
